src/gui/render.cpp

```diff
@@ -549,30 +549,44 @@ static bool force_square_pixels     = false;
 static bool force_vga_single_scan   = false;
 static bool force_no_pixel_doubling = false;
 
-// We double-scan VGA modes and pixel-double all video modes by default unless:
+// Double-scan VGA modes and pixel-double all video modes by default unless:
 //
-//  1) Single-scanning or no pixel-doubling is forced by the OpenGL shader.
-//  2) The interpolation mode is nearest-neighbour.
+//  1) Single-scanning or no pixel-doubling is requested by the OpenGL shader.
+//  2) The interpolation mode is nearest-neighbour in texture output mode.
 //
-// About the first point: the default `interpolation/sharp.glsl` shader forces
-// both because it scales pixels as flat adjacent rectangles. This not only
-// produces identical output versus double-scanning and pixel-doubling, but
-// also provides finer integer scaling steps (especially important on sub-4K
-// screens) and improves performance on low-end systems like the Raspberry Pi.
+// The default `interpolation/sharp.glsl` shader requests both single-scanning
+// and no pixel-doubling because it scales pixels as flat adjacent rectangles.
+// This not only produces identical output versus double-scanning and
+// pixel-doubling, but also provides finer integer scaling steps (especially
+// important on sub-4K screens), plus improves performance on low-end systems
+// like the Raspberry Pi.
+//
+// The same reasoning applies to nearest-neighbour interpolation in texture
+// output mode.
 //
 static void setup_scan_and_pixel_doubling()
 {
-	const auto nearest_neighbour_enabled = (GFX_GetInterpolationMode() ==
-	                                        InterpolationMode::NearestNeighbour);
+	const auto nearest_neighbour_on = (GFX_GetInterpolationMode() ==
+	                                   InterpolationMode::NearestNeighbour);
 
-	force_vga_single_scan   = nearest_neighbour_enabled;
-	force_no_pixel_doubling = nearest_neighbour_enabled;
+	switch (GFX_GetRenderingBackend()) {
+	case RenderingBackend::Texture:
+		force_vga_single_scan   = nearest_neighbour_on;
+		force_no_pixel_doubling = nearest_neighbour_on;
+		break;
 
-	if (GFX_GetRenderingBackend() == RenderingBackend::OpenGl) {
+	case RenderingBackend::OpenGl: {
 		const auto shader_info = get_shader_manager().GetCurrentShaderInfo();
+		const auto none_shader_active = (shader_info.name == NoneShaderName);
+
+		force_vga_single_scan = nearest_neighbour_on && none_shader_active;
+		force_no_pixel_doubling = nearest_neighbour_on && none_shader_active;
 
 		force_vga_single_scan |= shader_info.settings.force_single_scan;
 		force_no_pixel_doubling |= shader_info.settings.force_no_pixel_doubling;
```

Comment:
This looks a little strange that we're doing a bitwise or on a boolean value.  It's still technically correct though so you can leave it if you want.  That's the only comment I have.  Looks like a pretty straightforward fix otherwise :+1:

Reply:
Yeah I'll change it as @kcgen suggested.